log.go

```diff
@@ -47,6 +47,12 @@ type Log struct {
 
 	// Data holds the log entry's type-specific data.
 	Data []byte
+
+	// Extensions holds an opaque byte slice of information for middleware. It
+	// is up to the client of the library to properly modify this as it adds
+	// layers and remove those layers when appropriate. This value is a part of
+	// the log, so very large values could cause timing issues.
+	Extensions []byte
```

Comment:
I'd like to test this for backward compatibility. I think it should be OK since msgpack should be able to decode old messages into this struct and new messages into the old struct although it will strip the Extensions field in the process.

For chunking old servers stripping is probably not a big deal because they will crash when they try to decode/apply the partial chunk anyway and so upgrade will need to be managed by client or operator somehow.

But in general, I'm not sure if we need to make this change more explicit somehow. The way we've dealt with compatibilities in raft before is via protocol numbers which are negotiated by raft peers at start and can be forced to older version to ensure compatibility through an upgrade. That is fairly heavyweight though e.g. in Consul requires two rolling restarts to safely upgrade servers through a raft protocol change when we could just automatically disallow using the new chunking feature until all servers are running a new version.

We should probably _at least_ document in this comment that setting this is only safe when all raft peers are known to be running a version of raft with this change in though? Violating the assumption that log playback on all replicas will behave the same seems risky in general here. Imagine another extension that isn't chunking and doesn't result in invalid payloads but just alters how the FSM should apply the log - in that case failure of the client system to manage versioning manually would cause silent FSM corruption and divergence between replicas.

Reply:
Devils advocate says that it's the library client's responsibility to ensure that FSMs are backwards compatible on top of this library. Since this doesn't actually change the encoding format or cause anything to inherently break between peers other than the interpretation of the log messages themselves, it can stay like it is and we should just document (somehow?) that clients need to manage compatibility concerns themselves.

This is all assuming that this is actually backwards compatible and will be decoded by an old client from a log or RPC without crashing etc. I think that's the case but we should test it to be sure before accepting.

Reply:
I'll add some backwards compat testing in the chunking library. And I'll add some documentation as you suggested.

As for the upgrade path I also took the position that it's up to the client. So long as you don't trigger an extension -- in this case, chunking, by going above the limit (which is the same as it was before) -- then you don't run the risk of triggering chunking. So for any scenario in which we tell someone to turn on larger value support in Consul, we'd also tell them that all Consul nodes _must_ be at least version X. And if we ever publicly document that option, we would do the same thing, e.g. warning, do not make higher than the default if all Consul servers are not at least version X.

Reply:
I've added an upgrade warning here. I'll work on the backwards compat tests shortly.

Reply:
I did some testing with Consul, one node using raft including Extensions and one without. As I think you suspected, for a normal sized apply that didn't trigger ApplyLog, everything worked fine. When I went through ApplyLog it caused a decoding panic on the Consul node prior to this PR. (Technically it was an error; Consul just panics in this situation).